### ISSUE TITLE
Separate design and functional spec?

### ISSUE THREAD

**Issue by faulda** [HighSchoolHacking/GLS#68]
Functional spec: usually specifies details about the functions that a part of the system must perform. Contains information about inputs the component requires and output it produces, as well as defining expected behavior (handling of success and error cases). It does not include the *how*, the underlying code that makes it work. For a new GLS command, this would include:  
- syntax of a new command, plus examples
- output of the command in the various languages, plus examples
- 1 or 2 use cases where the new command would be used
- scoping information (for example, features that will be locked down because 1 or more languages don't support it, like multiline lambdas or default member variable values)

Design spec: usually specifies details about *how* the system accomplishes the behavior specified in the functional spec. For a new GLS command, this would include:
- new language properties that must be added
- general description of command implementation logic
- basic description of test coverage (this could also be separated into a test spec, but that might be a bit overkill)

So the question is, do we want to create separate documents like this?
I don't want to bury us in documentation and paperwork to the point where it slows down code output. On the other hand, the things listed above would be good to write down and discuss before implementing features so that major design decisions are made upfront and not later on-the-fly.

**Comment by JoshuaKGoldberg** [comment]
I don't think we need functional specs for commands They would certainly be useful for the overall GLS program structure, but each command is a very small amount of logic.

Having design docs on relevant *groups* of commands such as class declarations or imports (a group defined as >=1 similar commands) the way we've been doing seems to me like the most useful to an outside reader.

Agreed that individual test specs are overkill for this. There should be a functional spec for the testing framework itself, but not the individual tests. Those are self-explanatory.

**Comment by faulda** [comment]
Sounds good. If we do include both syntax/functional stuff and properties/design stuff in the docs, we should at least put them in separate sections I think.
